File: app/src/main/java/com/alva/testbrowser/fragment/BookMarkFragment.java
package com.alva.testbrowser.fragment;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.alva.testbrowser.adapter.BookmarkAdapter;
import com.alva.testbrowser.R;
import com.alva.testbrowser.database.Bookmark;
import com.alva.testbrowser.database.MyDatabase;
import com.alva.testbrowser.util.RecordViewModel;

import java.util.ArrayList;
import java.util.List;

public class BookMarkFragment extends Fragment {
    MyDatabase dataBase;
    RecordViewModel recordViewModel;
    BookmarkAdapter adapter;
    private List<Bookmark> bookmarkList = new ArrayList<>();

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.bookmark_layout, container, false);
        ListView listView = view.findViewById(R.id.bookmark_view);
        ImageButton button = view.findViewById(R.id.bookmarkDelete);

        this.recordViewModel = new ViewModelProvider.AndroidViewModelFactory(getActivity().getApplication()).create(RecordViewModel.class);
        this.recordViewModel.getAllBookmarks();
        //BookmarkAdapter1 adapter = new BookmarkAdapter1(getActivity(), R.layout.simple_adapter_item, this.recordViewModel.bookmarkList);
        adapter = new BookmarkAdapter(getActivity(), R.layout.simple_adapter_item, recordViewModel.bookmarkList);
        listView.setAdapter(adapter);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Bookmark bookmark = recordViewModel.bookmarkList.get(position);
                getActivity().setResult(Activity.RESULT_OK, new Intent().putExtra("open_url", bookmark.getUrl()));
                getActivity().finish();
            }
        });

        //长按监听
        listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                Bookmark bookmark = recordViewModel.bookmarkList.get(position);
                Log.d("longClick", bookmark.getTitle());
                //弹出菜单
                PopupMenu popupMenu = new PopupMenu(getActivity(), view);
                popupMenu.getMenuInflater().inflate(R.menu.menu_bookmark, popupMenu.getMenu());
                popupMenu.show();
                popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem menuItem) {
                        switch (menuItem.getItemId()) {
                            case R.id.editItem:
                                optionEdit(position, listView);
                                break;
                            case R.id.deleteItem:
                                optionDelete(position, listView);
                                break;
                            default:
                                throw new IllegalStateException("Unexpected value: " + menuItem.getItemId());
                        }
                        return true;
                    }
                });
                return true;
            }
        });

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder dialogbuilder = new AlertDialog.Builder(getActivity());
                dialogbuilder.setTitle("DELETE");
                dialogbuilder.setMessage("确认删除所有书签吗");
                dialogbuilder.setCancelable(false);
                dialogbuilder.setPositiveButton("确认", new DialogInterface.OnClickListener() {
                    //按了确定后删除数据，并且刷新列表
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        recordViewModel.deleteAllBookmarks();
                        adapter.deleteAllBookmark();
                        adapter.notifyDataSetChanged();
                    }
                });
                dialogbuilder.setNegativeButton("取消", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                    }
                });
                AlertDialog dialog = dialogbuilder.create();
                dialog.show();
                dialog.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(Color.BLACK);
                dialog.getButton(DialogInterface.BUTTON_NEGATIVE).setTextColor(Color.BLACK);
            }
        });
        return view;
    }

    public void optionDelete(int position, ListView listView) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("DELETE");
        builder.setMessage("是否删除此书签？");
        builder.setCancelable(false);
        builder.setPositiveButton("确认", new DialogInterface.OnClickListener() {
            //按了确定后删除数据，并且刷新列表
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Bookmark bookmark = recordViewModel.bookmarkList.get(position);
//                Log.d("output_title",bookmark.getTitle());
                recordViewModel.deleteBookmark(bookmark);
//                Log.d("delete_url",bookmark.getUrl());
                recordViewModel.getAllBookmarks();
                //从列表中删除
                adapter.deleBookmarkItem(position);
//                Log.d("dele_length", String.valueOf(recordViewModel.bookmarkList.size()));
            }
        });
        builder.setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
        ((AlertDialog) dialog).getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(Color.BLUE);
        ((AlertDialog) dialog).getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(Color.BLUE);
    }

    public void optionEdit(int position, ListView listView) {
        final EditText editText = new EditText(getActivity());
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity(), 3);
        builder.setTitle("请输入新标题");
        builder.setView(editText);
        editText.setText(recordViewModel.bookmarkList.get(position).getTitle());
        //单点全选
        editText.setSelectAllOnFocus(true);
        builder.setPositiveButton("确定", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                String title = editText.getText().toString();
                adapter.updateBookmarkItem(position, title);
                Bookmark bookmark = recordViewModel.bookmarkList.get(position);
                bookmark.setTitle(title);
                recordViewModel.updateBookmark(bookmark);
            }
        });
        builder.setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });
        builder.create().show();
    }

}

File: app/src/main/java/com/alva/testbrowser/BookmarkFragment.kt
package com.alva.testbrowser

import android.app.AlertDialog
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.alva.testbrowser.Adapter.BookmarkAdapter
import com.alva.testbrowser.databinding.DialogEditWebBinding
import com.alva.testbrowser.databinding.FragmentBookmarkBinding
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class BookmarkFragment : Fragment() {
    private var _binding: FragmentBookmarkBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding = FragmentBookmarkBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val viewModel by activityViewModels<BookmarkViewModel>()
        binding.recyclerView.layoutManager = LinearLayoutManager(requireContext())
        viewModel.allWebsLive.observe(viewLifecycleOwner, {
            binding.recyclerView.adapter = BookmarkAdapter(viewModel)
        })
        binding.deleteButton.setOnClickListener {
            val builder: AlertDialog = AlertDialog.Builder(requireContext())
                .setTitle(R.string.dialog_delete_bookmark_title)
                .setPositiveButton(R.string.dialog_positive_message) { _, _ ->
                    viewModel.deleteAllWebs()
                }
                .setNegativeButton(R.string.dialog_negative_message) { dialog, _ ->
                    dialog.cancel()
                }
                .show()
            builder.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(Color.BLACK)
            builder.getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(Color.BLACK)
        }
        binding.addButton.setOnClickListener {
            val v = View.inflate(it.context, R.layout.dialog_edit_web, null)
            val dialogBinding = DialogEditWebBinding.bind(v)
            val builder: AlertDialog =
                AlertDialog.Builder(it.context)
                    .setTitle(R.string.dialog_add_title)
                    .setView(v)
                    .setPositiveButton(R.string.dialog_positive_message) { _, _ ->
                        val web = Bookmark(
                            dialogBinding.editTextName.text.toString(),
                            dialogBinding.editTextUrl.text.toString()
                        )
                        viewModel.insertWebs(web)
                    }
                    .setNegativeButton(R.string.dialog_negative_message) { dialog, _ ->
                        dialog.cancel()
                    }
                    .show()
            builder.getButton(AlertDialog.BUTTON_POSITIVE).apply {
                setTextColor(Color.GRAY)
                isEnabled = false
            }
            builder.getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(Color.BLACK)
            dialogBinding.editTextName.requestFocus()
            lifecycleScope.launch {
                delay(100)
                (requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).showSoftInput(
                    dialogBinding.editTextName,
                    0
                )
            }
            dialogBinding.editTextUrl.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence?,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    if (s.toString().isNotEmpty()) {
                        builder.getButton(AlertDialog.BUTTON_POSITIVE).apply {
                            setTextColor(Color.BLACK)
                            isEnabled = true
                        }
                    } else {
                        builder.getButton(AlertDialog.BUTTON_POSITIVE).apply {
                            setTextColor(Color.GRAY)
                            isEnabled = false
                        }
                    }
                }

                override fun afterTextChanged(s: Editable?) {}
            })
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}